[MRG+1] Add OneVs{One,All}Classifier._pairwise: fix for #7306

rsmith54 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

Fixes #7306.
#### What does this implement/fix? Explain your changes.

Adds `_pairwise` to `OneVsOneClassifier` and `OneVsAllClassifier`  and adds a test to check this is properly set.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
fix for #7306

amueller commented on sklearn/multiclass.py:
```diff
@@ -391,6 +391,11 @@ def intercept_(self):
 
+    @property
+    def _pairwise(self):
+        '''Indicate if wrapped estimator is using a precomputed Gram matrix'''
```
Use double quotes ``"""`` please.

Commit:
fix to use double quotes

amueller commented on sklearn/tests/test_multiclass.py:
```diff
@@ -605,3 +605,13 @@ def test_ecoc_gridsearch():
     best_C = cv.best_estimator_.estimators_[0].C
     assert_true(best_C in Cs)
+
+def test_pairwise_attribute():
```
This test is good though I'd prefer if there was (additionally?) a test that checks that the behavior is actually the one we want. Maybe check that ``cross_val_score`` works correctly? Also, you test only tests one of the two properties.

rsmith54 replied:
Do you mean add a test similar to the one here? 

https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/tests/test_cross_validation.py#L714

amueller replied:
yes

rsmith54 commented:
Just fixed the single to double quotes

amueller commented on sklearn/tests/test_multiclass.py:
```diff
@@ -605,3 +605,13 @@ def test_ecoc_gridsearch():
+    clf_precomputed = svm.SVC(kernel='precomputed')
+    clf_notprecomputed = svm.SVC()
+
+    ovrFalse = OneVsRestClassifier(clf_notprecomputed)
```
```python
for MultiClassClassifier in [OneVsRestClassifier, OneVsOneClassifier]:
```
?

Commit:
check both types of classifiers

amueller commented:
is it clear what I mean with the test for ``cross_val_score``?

rsmith54 commented:
Not exactly, we crossposted though.  Something similar to the test I linked to ?

Commit:
adding test with cross_val_score

rsmith54 commented:
Okay, so I added another test which checks the cross_val_score with precomputed vs linear kernels, but this only works for ```OneVsRestClassifier```, not ```OneVsOneClassifier```.  When running with a ```OneVsOneClassifier```, I get 

```nose.proxy.ValueError: X.shape[0] should be equal to X.shape[1]```, although printing out the shape they naively seem the same. Is this expected (it's possible I'm missing the reason) ?

jnothman commented on sklearn/tests/test_multiclass.py:
```diff
@@ -605,3 +611,33 @@ def test_ecoc_gridsearch():
+
+#    for MultiClassClassifier in [OneVsRestClassifier, OneVsOneClassifier]:
+    for MultiClassClassifier in [OneVsRestClassifier]:
+        ovrFalse = MultiClassClassifier(clf_notprecomputed)
```
please avoid camel case

amueller replied:
(unless it's a class name)

jnothman commented:
Not sure where you're saying tests fail: it looks like tests are passing except the PEP8 check, which may be merely due to your print statements.

rsmith54 commented:
They all succeed as is, but I have commented out 
```#    for MultiClassClassifier in [OneVsRestClassifier, OneVsOneClassifier]:```

The failure happens with the OneVsOneClassifier.

Commit:
removing camelcase and print statements

jnothman commented:
oh, right.

jnothman commented:
```
Running flake8 on the diff in the range 5305861..6aee368 (5 commit(s)):
--------------------------------------------------------------------------------
./sklearn/tests/test_multiclass.py:615:1: E302 expected 2 blank lines, found 1
def test_pairwise_attribute():
^
./sklearn/tests/test_multiclass.py:626:1: E302 expected 2 blank lines, found 1
def test_pairwise_cross_val_score():
^
./sklearn/tests/test_multiclass.py:641:9: F841 local variable 'clf' is assigned to but never used
        clf = clf_notprecomputed
        ^
```

regarding the error, could you leave the broken test in or else report the full traceback so we don't necessarily need to go run it to help you debug?

Commit:
check both OveVsRest and OneVsOne

rsmith54 commented:
Hi, yes, I have reincluded it.

Commit:
fix for PEP8

jnothman commented on sklearn/tests/test_multiclass.py:
```diff
@@ -605,3 +611,33 @@ def test_ecoc_gridsearch():
+
+    for MultiClassClassifier in [OneVsRestClassifier, OneVsOneClassifier]:
+        ovrFalse = MultiClassClassifier(clf_notprecomputed)
+        assert_false(ovrFalse._pairwise)
```
camelCase

jnothman commented on sklearn/tests/test_multiclass.py:
```diff
@@ -30,6 +32,10 @@
 
+with warnings.catch_warnings():
+    warnings.simplefilter('ignore')
+    from sklearn import cross_validation as cval
```
Please us `sklearn.model_selection` instead.

jnothman commented:
Right. I've never touched the 1v1 code before. [Here](https://github.com/scikit-learn/scikit-learn/blob/5305861/sklearn/multiclass.py#L403) 1v1 pulls out only some samples from X without respect to `_pairwise`. It should be using `_safe_split`. We should probably move `_safe_split` to utils and use it there in 1v1.

Commit:
use cross_val_score from model_selection

rsmith54 commented:
I tried to play around with _safe_split, but I'm not exactly sure how to change that code to use it properly.  Any hints? Thanks!

jnothman commented:
https://github.com/scikit-learn/scikit-learn/blob/5305861/sklearn/multiclass.py#L403 should be changed to be something like
```python
return _fit_binary(estimator, _safe_split(estimator, X, y, indices=ind[cond])[0], y_binary, classes=[i, j])
```

ideally we wouldn't duplicate work with `y`, but that's minor.

Commit:
try with _safe_split

rsmith54 commented:
Okay, so I had tried something similar (at least it gave me the same error as we now see.  It breaks a bunch of the other tests as well, always in these lines :
```
  File "scikit-learn/sklearn/cross_validation.py", line 1638, in _safe_split
    y_subset = safe_indexing(y, indices)
  File "scikit-learn/sklearn/utils/__init__.py", line 110, in safe_indexing
    return X.take(indices, axis=0)
```

jnothman commented:
error type and message?

rsmith54 commented:
Sorry cut it out, see above.

jnothman commented on sklearn/multiclass.py:
```diff
@@ -400,8 +406,7 @@ def _fit_ovo_binary(estimator, X, y, i, j):
     ind = np.arange(X.shape[0])
-    return _fit_binary(estimator, X[ind[cond]], y_binary, classes=[i, j])
-
+    return _fit_binary(estimator, _safe_split(estimator, X, y, indices=ind[cond])[0], y_binary, classes=[i, j])
```
sorry, my mistake. you shouldn't be passing y into `_safe_split`. Rather, `_safe_split` should probably allow `y=None`.

Commit:
use None with y

rsmith54 commented:
Okay, so that fixes the other tests, but our new test fails, again only on the ```OneVsOneClassifier```, even with _safe_split :

```ValueError: X.shape[1] = 99 should be equal to 66, the number of samples at training time
```

rsmith54 commented:
I guess looking at it a bit more that we need to add the same ```_safe_split``` call in the prediction logic somewhere also, but I'm not sure where.

jnothman commented:
By not moving `_safe_split` to `sklearn.utils`, you've created a circular dependency between `multiclass` and `cross_validation` modules.

Commit:
moving _safe_split to utils

jnothman commented:
I don't think it belongs in `utils.multiclass`. It might be worth just sticking it in `__init__` for now until someone comes up with something better.

jnothman commented:
Arguably, `utils.metaestimators` might be appropriate.

Recall, however, that `sklearn.cross_validation` is deprecated. You should be moving this from `model_selection/_validation.py`, and ideally don't touch `cross_validation` unless we decide to back-port the fix to the deprecated version.

rsmith54 commented:
Yeah, okay that makes a lot of sense.  I'm happy to move everything over to ```model_selection/_validation.py```.  I'm still not exactly sure where the fix should be to properly use ```_safe_split``` in the prediction logic.  Any clues where that should go?

jnothman commented:
I don't think it's relevant to the prediction logic.

On 13 September 2016 at 13:11, Russell Smith <notifications@github.com>
wrote:

> Yeah, okay that makes a lot of sense. I'm happy to move everything over to
> model_selection/_validation.py. I'm still not exactly sure where the fix
> should be to properly use _safe_split in the prediction logic. Any clues
> where that should go?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7350#issuecomment-246561022>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz67ewDPPAAs354Na_EM6mINxjtguCks5qphRJgaJpZM4J2L1m>
> .
>

Commit:
move to _safe_split to utils/metaestimators.py

Commit:
don't touch cross_validation and move from model_selection to utils instead

rsmith54 commented:
Okay, so I have reverted ```cross_validation.py``` and moved ```_safe_split``` from ```model_selection/_split.py``` to ```utils/metaestimators.py``` instead.

I still have the issue with the shape :
```
======================================================================
ERROR: sklearn.tests.test_multiclass.test_pairwise_cross_val_score
----------------------------------------------------------------------
Traceback (most recent call last):
  File "venv/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "scikit-learn/sklearn/tests/test_multiclass.py", line 637, in test_pairwise_cross_val_score
    score_precomputed = cross_val_score(ovr_true, linear_kernel, y)
  File "scikit-learn/sklearn/model_selection/_validation.py", line 168, in cross_val_score
    for train, test in cv.split(X, y, labels))
  File "scikit-learn/sklearn/externals/joblib/parallel.py", line 758, in __call__
    while self.dispatch_one_batch(iterator):
  File "scikit-learn/sklearn/externals/joblib/parallel.py", line 608, in dispatch_one_batch
    self._dispatch(tasks)
  File "scikit-learn/sklearn/externals/joblib/parallel.py", line 571, in _dispatch
    job = self._backend.apply_async(batch, callback=cb)
  File "scikit-learn/sklearn/externals/joblib/_parallel_backends.py", line 109, in apply_async
    result = ImmediateResult(func)
  File "scikit-learn/sklearn/externals/joblib/_parallel_backends.py", line 322, in __init__
    self.results = batch()
  File "scikit-learn/sklearn/externals/joblib/parallel.py", line 131, in __call__
    return [func(*args, **kwargs) for func, args, kwargs in self.items]
  File "scikit-learn/sklearn/externals/joblib/parallel.py", line 131, in <listcomp>
    return [func(*args, **kwargs) for func, args, kwargs in self.items]
  File "scikit-learn/sklearn/model_selection/_validation.py", line 280, in _fit_and_score
    test_score = _score(estimator, X_test, y_test, scorer)
  File "scikit-learn/sklearn/model_selection/_validation.py", line 304, in _score
    score = scorer(estimator, X_test, y_test)
  File "scikit-learn/sklearn/metrics/scorer.py", line 204, in _passthrough_scorer
    return estimator.score(*args, **kwargs)
  File "scikit-learn/sklearn/base.py", line 326, in score
    return accuracy_score(y, self.predict(X), sample_weight=sample_weight)
  File "scikit-learn/sklearn/multiclass.py", line 548, in predict
    Y = self.decision_function(X)
  File "scikit-learn/sklearn/multiclass.py", line 569, in decision_function
    predictions = np.vstack([est.predict(X) for est in self.estimators_]).T
  File "scikit-learn/sklearn/multiclass.py", line 569, in <listcomp>
    predictions = np.vstack([est.predict(X) for est in self.estimators_]).T
  File "scikit-learn/sklearn/svm/base.py", line 573, in predict
    y = super(BaseSVC, self).predict(X)
  File "scikit-learn/sklearn/svm/base.py", line 310, in predict
    X = self._validate_for_predict(X)
  File "scikit-learn/sklearn/svm/base.py", line 475, in _validate_for_predict
    (X.shape[1], self.shape_fit_[0]))
ValueError: X.shape[1] = 99 should be equal to 66, the number of samples at training time

----------------------------------------------------------------------
Ran 34 tests in 1.108s
```

jnothman commented:
Ah, you're right, we do need changes at predict time. Sorry.

In the kernelised version, you need to keep track of which indices (columns of `X`) were used to train each estimator, and make similar slices at predict time. So:
```
        Y = np.array([_predict_binary(e, X) for e in self.estimators_]).T
```

will look like:

```
        if self.pairwise_indices_ is None:
            Xs = [X] * len(self.estimators_)
        else:
            Xs = (X[:, idx] for idx in self.pairwise_indices_)
        Y = np.array([_predict_binary(e, X) for e in zip(self.estimators_, Xs)]).T
```

I'll leave it to you to set  `pairwise_indices_` as appropriate, to document and test this feature.

Commit:
getting out the conditional indices

Commit:
reverting to CamelCase for class name

Commit:
use list to loop twice

Commit:
removing prints

rsmith54 commented:
So I have gotten the new test to pass, by passing back the conditional indices from the ```_fit_ovo_binary``` back to the ```OneVsOneClassifier```.  I think there is some cleanup that could be done, and I still need to do the documentation, but is this an acceptable way to pass around/save the pairwise indices?

I guess I should also implement this for the partial fitting, right?  Thanks for all the help.

amueller commented:
hm for some reason travis is not run. Could you please rebase?

amueller commented:
LGTM. can you add an entry to whatsnew? I'd like to see the CI run, too

amueller commented:
partial fit could be much more tricky, but I'm not sure...

amueller commented:
something went wrong with that rebase... make sure to pull from master before

Commit:
fixing split test

Commit:
whatsnew entry

amueller commented:
great :) feel free to cleanup if you feel anything needs cleanup but this seems to be the right fix for the issue.

Commit:
moving GPKernel import

Commit:
fixing other tests

rsmith54 commented:
There was a small problem with the one of the tests from one of the imports, but hopefully now things should be okay.  Let me know if there is anything else I can do.

jnothman commented on doc/whats_new.rst:
```diff
@@ -456,6 +456,12 @@ Bug fixes
       by `Yichuan Liu <https://github.com/yl565>`_.
 
+      - Add ``_pairwise`` to :class:`OneVsOneClassifier` and
+	:class:`OneVsRestClassifier`.  Also moves ``_safe_split``
```
Hanging indentation should align this text with the `A` above.

jnothman commented:
`sklearn/model_selection/tests/test_split.py:def test_safe_split_with_precomputed_kernel` should be moved to `utils/tests/test_multiclass.py`

jnothman commented on sklearn/multiclass.py:
```diff
@@ -559,15 +572,27 @@ def decision_function(self, X):
+            Xs = [X[:, idx] for idx in indices]
+
+        predictions = np.vstack([est.predict(Xi) for est,Xi in zip(self.estimators_, Xs)]).T
+        confidences = np.vstack([_predict_binary(est, Xi) for est,Xi in zip(self.estimators_, Xs)]).T
```
PEP8

jnothman commented on sklearn/multiclass.py:
```diff
@@ -472,10 +480,14 @@ def fit(self, X, y):
-            delayed(_fit_ovo_binary)(
-                self.estimator, X, y, self.classes_[i], self.classes_[j])
-            for i in range(n_classes) for j in range(i + 1, n_classes))
+        estimators_indices = Parallel(n_jobs=self.n_jobs)(delayed(_fit_ovo_binary)(self.estimator, X, y,
```
PEP8

Commit:
moving test_safe_split_with_precomputed_kernel

jnothman commented on sklearn/multiclass.py:
```diff
@@ -399,9 +407,8 @@ def _fit_ovo_binary(estimator, X, y, i, j):
-    return _fit_binary(estimator, X[ind[cond]], y_binary, classes=[i, j])
-
+    indcond = np.arange(X.shape[0])[cond]
+    return _fit_binary(estimator, _safe_split(estimator, X, None, indices=indcond)[0], y_binary, classes=[i, j]) , indcond
```
PEP8

jnothman reviewed (changes requested):
Apart from these minor things, LGTM!

Commit:
final fixes to whats_new and pep8

rsmith54 commented:
I think pep8 problems should be fixed.  Let me know if there is anything else to change.

jnothman commented on sklearn/multiclass.py:
```diff
@@ -509,16 +525,19 @@ def partial_fit(self, X, y, classes=None):
+                delayed(_partial_fit_ovo_binary)(
+                    estimator, X, y, self.classes_[i], self.classes_[j])
+                for estimator, (i, j) in izip(
+                        self.estimators_, ((i, j)
```
For cosmetic reasons, can you pull the latter generator expression out as a separate variable `combinations`?

In fact, it can be set as `combinations = itertools.combinations(range(self.n_classes), 2)`

jnothman replied:
I know it's not really something you've changed in your code... but if you're going around cleaning things up, let's clean them up properly :)

jnothman commented:
You've introduced an error somewhere.

jnothman commented on sklearn/tests/test_multiclass.py:
```diff
@@ -24,7 +27,7 @@
                                   Perceptron, LogisticRegression)
 from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor
-from sklearn.model_selection import GridSearchCV
+from sklearn.model_selection import GridSearchCV,cross_val_score
```
PEP8

Commit:
fix import error and pep8

Commit:
fixing utils/tests/test_multiclass

Commit:
use itertools

jnothman commented:
We appear to have a circular dependency. Our options include:
* move `_ovr_decision_function` to `sklearn.utils.multiclass`
* do not (globally) import `GPKernel` for use in `_safe_split`. We can import it locally, and perhaps only if `'sklearn.gaussian_process' in sys.modules`
* both of the above

There are still flake8 issues listed at https://travis-ci.org/scikit-learn/scikit-learn/jobs/160044685

Commit:
fixes for pep8, local import for GPKernel, and moving ovr_decision_function to utils

Commit:
pep8

jnothman commented:
LGTM

jnothman commented:
Actually, a couple more minor things related to what you moved out of `sklearn/model_selection/tests/test_split.py`. `iris` there is no longer used. I'm not sure if `X` is used any longer, but it's pretty easy to test by removing it. In its new home, these may be better off as local variables.

jnothman commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -28,6 +28,19 @@
+perm = rng.permutation(iris.target.size)
+iris.data = iris.data[perm]
+iris.target = iris.target[perm]
+n_classes = 3
```
This is unused

jnothman commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -28,6 +28,19 @@
+from sklearn import datasets
+
+iris = datasets.load_iris()
+rng = np.random.RandomState(0)
```
This is unnecessary

jnothman commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -28,6 +28,19 @@
+from sklearn.svm import SVC
+from sklearn import datasets
+
+iris = datasets.load_iris()
```
Make this local to the test

Commit:
use local vars for test_safe_split_with_precomputed_kernel

jnothman commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -361,9 +349,14 @@ def test_class_distribution():
 
 
 def test_safe_split_with_precomputed_kernel():
+    from sklearn.model_selection import ShuffleSplit
```
Not local imports. That's very unconventional in tests. But it's useful only to load iris here.

Commit:
nonlocal imports

jnothman commented:
You still have unused `iris` in `test_split.py`

Commit:
remove iris from test_split

jnothman commented:
Now that we've spammed the CIs, thanks. Wait for @amueller to give this another pass.

jnothman commented:
@amueller, changes in the last 7 hours:
* `test_safe_split_with_precomputed_kernel` now lives alongside `_safe_split`
* @rsmith54 took the "both" option from https://github.com/scikit-learn/scikit-learn/pull/7350#issuecomment-247203533
* PEP8

amueller commented:
@jnothman I'll look tomorrow (though I also should get started on that keynote). Ugh the github review "feature" breaks the threading in email notifications, the only thing that I could use to keep at least try to keep track :(

amueller commented:
what do you use to keep track of the issue tracker?

jnothman commented:
> what do you use to keep track of the issue tracker?

Email. Not always easy. I haven't noticed the threading breakage.

raghavrv commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -345,3 +346,27 @@ def test_class_distribution():
+def test_safe_split_with_precomputed_kernel():
+    from sklearn.model_selection import ShuffleSplit
+    from sklearn.svm import SVC
+    from sklearn import datasets
```
Why are these imported here and not at the start of the file?

jnothman replied:
That's not in the latest version.

raghavrv replied:
Ah github confused me :@

jnothman commented on doc/whats_new.rst:
```diff
@@ -456,6 +456,12 @@ Bug fixes
       by `Yichuan Liu <https://github.com/yl565>`_.
 
+    - Add ``_pairwise`` to :class:`OneVsOneClassifier` and
+      :class:`OneVsRestClassifier`.  Also moves ``_safe_split``
```
It might read more like "Cross-validation of OvR and OvO models now works with precomputed kernels". No mention of `_safe_split`, as @raghavrv says.

rsmith54 replied:
Changed the ```whats_new``` to the suggestion from @jnothman .

Commit:
fixing whats_new

rsmith54 commented:
Is there anything else I need to do for this? Thanks for all the help!

jnothman commented:
Not at this stage, @rsivapr. @amueller or someone else may review this soon, but it might wait until after the 0.18.0 release. I'll label it 0.18 because it's so close to completion it might make it in, but I'd consider this a non-urgent bug fix.

jnothman commented:
If you're lucky, the next review will just be a double-check with no work from you.

ogrisel commented on sklearn/multiclass.py:
```diff
@@ -452,6 +463,7 @@ class OneVsOneClassifier(BaseEstimator, ClassifierMixin, MetaEstimatorMixin):
     def __init__(self, estimator, n_jobs=1):
         self.estimator = estimator
         self.n_jobs = n_jobs
+        self.pairwise_indices_ = None
```
The constructor of a scikit-learn estimators should never set attributes with a trailing `_`, it should only store hyperparameters as attributes. Attributes with a trailing `_` should only be set by the `fit` method or by a private submethod called only at `fit` time.

At test time, methods that need access to that attribute can check its presence with the `_check_fitted_model` helper.

ogrisel replied:
I thought we had a consistency check in `test_common` for this kind of things but maybe it's not applied on meta estimators.

jnothman replied:
Somehow I missed this, sorry.

ogrisel commented on sklearn/multiclass.py:
```diff
@@ -538,6 +556,7 @@ def predict(self, X):
         y : numpy array of shape [n_samples]
             Predicted multi-class targets.
         """
+
```
why the new line here?

ogrisel commented on sklearn/multiclass.py:
```diff
@@ -559,62 +578,29 @@ def decision_function(self, X):
+        if indices is None:
+            Xs = [X] * len(self.estimators_)
+        else:
+            Xs = [X[:, idx] for idx in indices]
```
Is this case tested? If not please add a dedicated test in `test_multiclass.py` with `SVC(kernel='precomputed')` and check the expected shape of the output.

rsmith54 replied:
Let me know if the test ```test_pairwise_indices``` is what you are looking for here.

ogrisel reviewed (changes requested):
Apart from the following comments, this LGTM.

jnothman commented on doc/whats_new.rst:
```diff
@@ -456,6 +456,11 @@ Bug fixes
 
+    - Cross-validation of :class:`OneVsOneClassifier` and
+      :class:`OneVsRestClassifier` now works with precomputed kernels.
+    (`#7350 <https://github.com/scikit-learn/scikit-learn/pull/7350/>`_)
```
Indentation

jnothman commented on sklearn/utils/tests/test_multiclass.py:
```diff
@@ -345,3 +350,23 @@ def test_class_distribution():
+    K = np.dot(X, X.T)
+
+    cv = ShuffleSplit(test_size=0.25, random_state=0)
+    tr, te = list(cv.split(X))[0]
```
Just to clarify, this is moved code, not new code. Still, it might be a good idea to improve it, as it's easy to do so.

jnothman commented on sklearn/utils/metaestimators.py:
```diff
@@ -77,3 +78,38 @@ def if_delegate_has_method(delegate):
+    else:
+        y_subset = None
+
+    return X_subset, y_subset
```
It is now used in `sklearn.multiclass` and `sklearn.model_selection`. Do you think it belongs in `model_selection`? Do you think it would be better off in `sklearn.utils.__init__` than `sklearn.utils.metaestimators`?

ogrisel replied:
`sklearn.model_selection._split` seemed like a good module to host a private helper function named `_safe_split`. But I don't care that much.

jnothman replied:
> sklearn.model_selection._split seemed like a good module to host a private helper function named _safe_split. But I don't care that much.

It was, but I think the name is not quite right. It's just harder to come up with a better one: `_pairwise_friendly_indexing`?

raghavrv commented:
> It is now used in sklearn.multiclass and sklearn.model_selection. Do you think it belongs in model_selection?

Even I feel this should reside inside `model_selection`... But as you pointed out the cyclic import error, I think we have to go with `sklearn.utils.__init__`?

jnothman commented:
> Even I feel this should reside inside model_selection... But as you pointed out the cyclic import error, I think we have to go with sklearn.utils.__init__?

I don't get why it should belong exclusively in model selection. It pertains to anything that indexes on `X` and has to respect the `_pairwise` attribute of a sub-estimator, i.e. any metaestimator with scissors.

Commit:
adding additional checks for multiclass utils

Commit:
fixing problem with pairwise_indices initialization

Commit:
adding test for pairwise indices shapes

Commit:
whats_new fix

raghavrv commented:
> I don't get why it should belong exclusively in model selection.

I think the naming, signature and the intended use are not that generic enough. At least a generic version of such a function should may be called `_safe_subset` and should not use the arg `training_indices`...

Commit:
fixes for pep8

rsmith54 commented:
I made the changes requested by @ogrisel , other than those related to ```model_selection``` vs ```metaestimators``` for ```safe_split```.

raghavrv commented on doc/whats_new.rst:
```diff
@@ -458,7 +458,7 @@ Bug fixes
       :class:`OneVsRestClassifier` now works with precomputed kernels.
-    (`#7350 <https://github.com/scikit-learn/scikit-learn/pull/7350/>`_)
+      (`#7350 <https://github.com/scikit-learn/scikit-learn/pull/7350/>`_)
       by `Russell Smith <https://github.com/rsmith54>`_.
```
You could also add your name to the bottom as we are expecting more amazing pull requests like these from you ;)

raghavrv replied:
Now that you have added it there you can remove the link here... :)

Commit:
adding name to bottom of whats_new

Commit:
no double link

ogrisel commented:
> I don't get why it should belong exclusively in model selection. It pertains to anything that indexes on X and has to respect the _pairwise attribute of a sub-estimator, i.e. any metaestimator with scissors.

As you wish. You can leave it where it is.

raghavrv commented:
Thanks @rsmith54!

jnothman commented:
This proved somewhat trickier than was first thought, so big thanks
@rsmith54!

On 20 September 2016 at 21:50, Raghav RV <notifications@github.com> wrote:

> Thanks @rsmith54 <https://github.com/rsmith54>!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7350#issuecomment-248279106>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz64ce5IWBCFJNC11mXMmiKX6YC-9Nks5qr8iAgaJpZM4J2L1m>
> .
>

rsmith54 commented:
Thanks for all the help!
